### PR TITLE
fix(Common): allow null/undefined values for `NgForTrackBy`

### LINKED ISSUE #13641: 2.4.1 breaks backward compatibility
Binding a an optional trackBy on ngFor gives;

"Track By must be a function but received undefined".

In previous versions, undefined was simply ignored, many PrimeNG users are having this problem now. Simply ignoring undefined instead of throwing an error should restore backward compatibility. We've thought breaking changes will be in major versions only, not with 2.x.

https://github.com/primefaces/primeng/issues/1652

### PR DESCRIPTION
**Pull request by vicb**
Reverts a breaking change introduced in 2.4.1 by #13420
fixes #13641

### CODE REVIEW

File: modules/@angular/common/src/directives/ng_for.ts [hunk 1]
```diff
@@ -91,9 +92,10 @@ export class NgFor implements DoCheck, OnChanges {
   @Input() ngForOf: any;
   @Input()
   set ngForTrackBy(fn: TrackByFn) {
-    if (typeof fn !== 'function') {
-      throw new Error(`trackBy must be a function, but received ${JSON.stringify(fn)}.
-      See https://angular.io/docs/ts/latest/api/common/index/NgFor-directive.html#!#change-propagation for more information.`);
+    if (fn != null && typeof fn !== 'function' && isDevMode()) {
```
Comment (DzmitryShylovich) [inline modules/@angular/common/src/directives/ng_for.ts]: isn't `typeof fn !== 'function'` already includes `fn != null`?

Comment (vicb) [inline modules/@angular/common/src/directives/ng_for.ts]: a function is not null but it is of type function

Comment (DzmitryShylovich) [inline modules/@angular/common/src/directives/ng_for.ts]: but `typeof null !== 'function'` is true and `typeof undefined !== 'function' is true so `fn != null` is redundant :)

Comment (DzmitryShylovich) [inline modules/@angular/common/src/directives/ng_for.ts]: plus I think `isDevMode()` should go first so if we are in prod mode we will avoid unnecessary chechks

File: modules/@angular/common/src/directives/ng_for.ts [hunk 2]
```diff
@@ -91,9 +92,10 @@ export class NgFor implements DoCheck, OnChanges {
...
+      this._console.warn(
```
Comment (mhevery) [inline modules/@angular/common/src/directives/ng_for.ts]: leave a comment to change `warn` to `new Error` in v4.0?

**Review by mhevery — APPROVED** [review]

File: modules/@angular/common/src/directives/ng_for.ts (hunk 1 above)
Comment (vicb) [inline modules/@angular/common/src/directives/ng_for.ts]: I think you're wrong try this logic with int, function and null/undefined and then try your logic.

moving `isDevMode` first could be a good idea, will update.

Comment (vicb) [inline modules/@angular/common/src/directives/ng_for.ts]: oh @DzmitryShylovich the whole point of this commit is not to error on null/undefined, that should be what you miss here.

**Comment by DzmitryShylovich** [comment]
@vicb 
> oh @DzmitryShylovich the whole point of this commit is not to error on null/undefined, that should be what you miss here.

then this patch doesn't make any sense. You can completely remove the check because angular will throw an error anyway. 
For example (from the test case):
```
<template ngFor let-item [ngForOf]="items" [ngForTrackBy]="5">{{ item }}</template>`
```
will throw `VM566 core.umd.js:3491 EXCEPTION: Error in ./App class App - inline template:2:10 caused by: this._trackByFn is not a function`.

The only situation we need to report is `trackBy: item?.Id` and in this case trackBy is equal to `null`.

**Comment by vicb** [comment]
@DzmitryShylovich 
- I was against merging your original PR because of the breaking change,
- The team decided otherwise and it got merged,
- It broke customers apps,
- I came with this PR to fix it,
- It was throwing an error at first,
- After some discussion I switched to console.warn,
- Might not be ideal but I think that's the best we can do.

**Comment by DzmitryShylovich** [comment]
it's ok just need to remove `fn != null` :)

**Comment by vicb** [comment]
> it's ok just need to remove fn != null :)

No. Let's say you have a component that take a tracking function as an `@Input` and is implemented with a `NgFor`, how would you do this ? Create and identity function when the input is `null/undefined` ?

Basically removing this check is saying that `let trackBy;` and `let trackBy = undefined;` have different behavior (undefined/null actually)

**Comment by DzmitryShylovich** [comment]
> Create and identity function when the input is null/undefined ?

yeah, `@Input() trackBy = (a, b) => a === b` 
It's inconvenient, but it's the only way to catch `trackBy: item?.Id` which is much more common mistake.

**Comment by DzmitryShylovich** [comment]
nvm I think we already did the job and people realized that they were wrong

**Pull request merged**